crates/store/re_chunk/src/helpers.rs

```diff
@@ -0,0 +1,360 @@
+use std::sync::Arc;
+
+use arrow2::array::Array as ArrowArray;
+
+use re_log_types::{TimeInt, Timeline};
+use re_types_core::{Component, ComponentName, SizeBytes};
+
+use crate::{Chunk, ChunkResult, RowId};
+
+// --- Helpers ---
+
+impl Chunk {
+    // --- Batch ---
+
+    /// Returns the raw data for the specified component.
+    ///
+    /// Returns an error if the row index is out of bounds.
+    #[inline]
+    pub fn component_batch_raw(
+        &self,
+        component_name: &ComponentName,
+        row_index: usize,
+    ) -> Option<ChunkResult<Box<dyn ArrowArray>>> {
+        self.components.get(component_name).map(|list_array| {
+            if list_array.len() > row_index {
+                Ok(list_array.value(row_index))
+            } else {
+                Err(crate::ChunkError::IndexOutOfBounds {
+                    kind: "row".to_owned(),
+                    len: list_array.len(),
+                    index: row_index,
+                })
+            }
+        })
+    }
+
+    /// Returns the deserialized data for the specified component.
+    ///
+    /// Returns an error if the data cannot be deserialized, or if the row index is out of bounds.
+    #[inline]
+    pub fn component_batch<C: Component>(&self, row_index: usize) -> Option<ChunkResult<Vec<C>>> {
+        let res = self.component_batch_raw(&C::name(), row_index)?;
+
+        let array = match res {
+            Ok(array) => array,
+            Err(err) => return Some(Err(err)),
+        };
+
+        let data = C::from_arrow(&*array);
+        Some(data.map_err(Into::into))
+    }
+
+    // --- Instance ---
+
+    /// Returns the raw data for the specified component at the given instance index.
+    ///
+    /// Returns an error if either the row index or instance index are out of bounds.
+    #[inline]
+    pub fn component_instance_raw(
+        &self,
+        component_name: &ComponentName,
+        row_index: usize,
+        instance_index: usize,
+    ) -> Option<ChunkResult<Box<dyn ArrowArray>>> {
```

Comment:
If an missing component array is semantically consider "empty", then a row_index within an empty array should be an error. I'm not sure we need an Option here at all.

Reply:
Like I commented above: I wanted these APIs to be a lower-level building block. It can be valuable to the caller to know whether the component as a whole is missing with a single API.

Higher-level APIs with the kind of semantics you're describing are introduced later in this PR stream at the `re_query` layer, and we can decide to treat the missing component as an error there.